### ISSUE TITLE
WP components audit: text input fields

### ISSUE THREAD

**Issue by davewhitley** [WordPress/gutenberg#16685]
**Is your feature request related to a problem? Please describe.**

This issue is part of a review on the naming, structure, and composition of the UI components as brought up in https://github.com/WordPress/gutenberg/issues/16367

These are the current components that are related to text input fields that I considered in this issue:

- Autocomplete
- BaseControl
- FormTokenField
- TextControl
- TextareaControl

# [TextControl](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/text-control) / [TextareaControl](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/textarea-control)

![image](https://user-images.githubusercontent.com/618551/61156362-d2dac080-a4b8-11e9-889d-2eceb07492b7.png)

## Audit
Opportunities I came across while reviewing this component.

- "TextControl" and "TextareaControl" were confusing names because I didn't understand what "control" meant. Not ideal names for someone looking for a simple text field component.
- "TextControl" could mean many different things.
- Two distinct components didn't feel necessary, and could be simplified.

## Suggestions

- Combine TextControl and TextareaControl into “**TextField**”. Because they are so similar, we can increase usability of the system by combining the two.
- **TextField** can have a `multiline` prop to enable a multiline text field. A `textarea` html element would be used in this case. This aligns with the [`aria-multiline` attribute](https://www.w3.org/TR/wai-aria-1.1/#aria-multiline).

# [FormTokenField](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/form-token-field)

![image](https://user-images.githubusercontent.com/618551/61157117-d2dbc000-a4ba-11e9-8d04-e70e3f65936b.png)

## Audit
Opportunities I came across while reviewing this component.

- Based on the name, it's seems to unnecessarily imply a grouping with form elements.
- There's an opportunity here to generalize the "token" element into more reusable components, and not require a text field.

## Suggestions

- Consider renaming token to "**chip**" because it's more memorable, and I think more descriptive. "Tag" would also be acceptable but obviously a conflict with common WP language.
- Break this component down into several components: **Chip**, **ChipSet** (or ChipGroup), and **ChipTextField** (previously FormTokenField).

![image](https://user-images.githubusercontent.com/618551/61157550-09fea100-a4bc-11e9-99db-e92c2439bd58.png)

# [BaseControl](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/base-control)

## Audit
I could use some feedback on this one. It took me a while to understand what it actually provided and how it relates to the input elements like Radio and Checkbox. I’ve learned that it provides a **label** and **help text** for some components. In the future, I'm sure it could provide validation and RTL handling perhaps. 

Different strategies I've found in other component libraries:

- MDC Material has [FormField](https://href.li/?https://github.com/material-components/material-components-web/tree/master/packages/mdc-form-field) which handles labels, and RTL for inputs, but no help text.
- Their React library, [MDC React](https://href.li/?https://github.com/material-components/material-components-web-react), does not provide FormField as a component ([intentionally it seems](https://href.li/?https://github.com/material-components/material-components-web-react/issues/929#issuecomment-502186555)). Things like help text are implemented as separate sub components as far as I can tell. For example, [TextField](https://href.li/?https://github.com/material-components/material-components-web-react/tree/master/packages/text-field) has its own TextFieldHelperText component.
- Material UI has [FormControl](https://href.li/?https://material-ui.com/api/form-control/#formcontrol-api) which handles help text as well as the label.
- Ant Design has [Form.Item](https://href.li/?https://material-ui.com/api/form-control/#formcontrol-api).
- [Evergreen](https://evergreen.segment.com/components/form-field) has FormField, which provides label, isRequired, description, hint, and validationMessage. Apparently it's already combined in their text input and select components.

These different strategies are interesting to me, but I don’t know why some libraries went the direction they chose, or what the pros/cons are. Whatever we decide to do, we should weigh the options and take into consideration that we need to add field validation to some of these components. Material seems to only provide validation on text input elements. Maybe that’s why they don’t provide a help text in their FormField? Do we really need validation on checkboxes and radios, etc.? Regardless, we should have a `helptext` prop on all applicable elements for ease of use.

## Suggestions

Aside from the features it provides, I think any of the names above are better alternatives for BaseControl (**FormField**, **FormItem**, **FormControl**, etc.)

# [Autocomplete](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/autocomplete)

I wouldn't list this as a top-level component. Because it seems like an enhanced TextField or a variation, I would group it with that.

# Feedback

The feedback I'm looking for is related to naming, structure, and composition. I'm not looking for visual feedback of the components. Prop audit can be seen in the [comment below](https://github.com/WordPress/gutenberg/issues/16685#issuecomment-513343858).

**Comment by drw158** [comment]
# Props audit

As an addendum, here is a deeper evaluation of the props for these components. This is an effort to expand the components to be more useful and to answer the [question](https://github.com/WordPress/gutenberg/issues/16367), "Are properties well thought out and consistently applied?" This only covers visual props. Event handler props will be evaluated at a later date.

## TextField

### Validation

Validation is complex. A simple strategy that many libraries employ is to just have a default non-error state and an error state. This means there is no "success" or "warning" state. I like the simplicity of this and seems easier to implement. It also mirrors what is available in native html5 validation. 

On the other hand, it can be useful to provide the user with encouraging feedback when the they correct the error with a "success" state.

We could start small and expand if we need a success state.

- **validationMessage** or **errorMessage**: the error message to display if the input is invalid.
- **variant**: if we need more states. Accepts _danger_, _success_, and _warning_.

### ✨ New
- **fullWidth**: 100% width of container 
- **multiline**: makes a multiline text field (textarea)
- **hideLabel**: to visually hide the label while retaining accessibility.
- **dense**: makes the field smaller, for denser layouts
- **prefix** and **suffix**: Text to display before/after input value
- **iconStart** and **iconEnd***: Icon located at the start/end of the input
- **characterCount***: display a character counter next to the field
- **required**
- **disabled**
- **readOnly**
- **placeholder**

### 📋 Rename
- **help** → helpText

### 👍 No change
- **label**
- **type**

---

**Comment by mapk** [comment]
I'm closing out the Component Audit issues for now.
